src/wallet/wallet.cpp

```diff
@@ -860,7 +862,16 @@ std::pair<PaymentAddress, RecipientType> CWallet::GetPaymentAddressForRecipient(
         const uint256& txid,
         const libzcash::RecipientAddress& recipient) const
 {
+    AssertLockHeld(cs_wallet);
+
     auto self = this;
+
+    auto nHeight = chainActive.Height();
```

Comment:
If you're using `chainActive` here, you need to also lock on `cs_main`. We should only hold that lock temporarily.

Reply:
Given that this is only used as a fallback for the "what is the default set of receiver types" heuristic, why do we need to lock? Here, we should be okay with dirty reads.

Reply:
`GetDepthInMainChain` contains `AssertLockHeld(cs_main)`, so we need it to be held there. And we need to ensure that `chainActive` and `GetDepthInMainChain` accesses are synchronized, otherwise we could end up with a negative height.